### ISSUE TITLE
Istiod consumes a huge amount of memory with Discovery Selectors specified

### ISSUE THREAD

**Issue by zhaohuabing** [istio/istio#41658]
### Bug Description

I'm using Istio to manage a dozen of k8s clusters, with thousands of pods and service in these clusters.  For each cluster, only one or two namespce need to be discovered by the Istiod so I set Discovery Selectors to select these namespace.

I expect that the memory usage would match the pods and services of the selected namespace, however, it turns out the istiod memory increases with the scale of clusters. In my test environment, istio eats 10 G memory after 10 clusters have been added. From the flame graph of istio heap, most of the memory is cosumed by pods cache.

![Screen Shot 2022-10-27 at 15 35 47](https://user-images.githubusercontent.com/1321762/198246141-e7314a67-1654-4318-b11f-60f4ab6244ad.png)

 

### Version

```prose
istioctl version
client version: 1.12.9
control plane version: 1.12.6

➜  ~ kubectl version --short
Client Version: v1.22.5
Server Version: v1.16.3-tke.15
WARNING: version difference between client (1.22) and server (1.16) exceeds the supported minor version skew of +/-1
```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [X] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### Is this the right place to submit this?

- [X] This is not a security vulnerability
- [X] This is not a question about how to use Istio

**Comment by zhaohuabing** [comment]
The cause of this issue ist that the [kube serviceregistry](https://github.com/istio/istio/blob/293b73a9e8a56cad7147ac3ab627d9b731829cb4/pilot/pkg/serviceregistry/kube/controller/controller.go#L376) watches all the pods and uses the discovery selectors at the client side. 

After applying the correct LabelSelector when createing the SharedInformerFactory, the memory downsized from 10 G to 1 G.

**Comment by dddddai** [comment]
It's tricky to do service-side selecting since discovery selectors can be modified dynamically, so we would have to recreate the informers...

**Comment by zhaohuabing** [comment]
Actually, once a discovery selector has been set, it don't need to be changed at runtime. Normally we can just modify the label of a namespace to add or remove it from a mesh.

**Comment by howardjohn** [comment]
Discovery selector is matching namespace labels. You can't query k8s for "All pods in namespace matching X label". Actually you can;t even query for anything other than "all namespaces" or "one namespace" without opening a bunch of separate watches

**Comment by howardjohn** [comment]
How many pods total to get to 10gb ram btw?

One thing we could consider doing is make a `StripUnusedFields` for pods

**Comment by zhaohuabing** [comment]
> Discovery selector is matching namespace labels. You can't query k8s for "All pods in namespace matching X label". Actually you can;t even query for anything other than "all namespaces" or "one namespace" without opening a bunch of separate watches

In my case, only the pods in the selected namespace need to be insert a sidecar, so I can use label selectors to do server side filtering on pods instead of namespaces.  If the Discovery Selector mechanism can also be applied to sidecar injection webhook, that would solve the problem.

**Comment by zhaohuabing** [comment]
> StripUnusedFields

@howardjohn how can I strip the unsed files?

**Comment by hzxuzhonghu** [comment]
Right, discoverySelector isonly applied to namespace，not service pod， etc. So definitely can not be set to informer options.

Wonder how much before, want to have a compare

**Comment by dddddai** [comment]
> how can I strip the unsed files?

You would have to find out all pod fields that istio is using, I have tried that before, please see https://github.com/dddddai/istio/commit/e36cc5de293abe512d157fefb27420b7d65d91a3

Could you test it out and see if it helps? If it does I can file a PR, I'm not 100% sure I didn't miss any fields though
